## 1. Making ranges

In [2]:
type(range(15))

range

In [3]:
x = list(range(15))
x

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]

In [5]:
L = [3,5,7,8]
for ind in range(len(L)):
    print(ind)

0
1
2
3


In [32]:
L = ['a','c']

for ind in range(len(L)):
    print(L[ind])

a
c


In [33]:
list(range(1,5))

[1, 2, 3, 4]

In [34]:
list(range(5,100))

[5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99]

In [11]:
list(range(5,100,5))

[5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]

In [6]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [7]:
list(range(10,-1,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

In [35]:
type(x)

list

In [6]:
L = list(range(10,-1,-1))
print(L)
L[2:8]

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


[8, 7, 6, 5, 4, 3]

In [7]:
L[8:2:-1]

[2, 3, 4, 5, 6, 7]

In [37]:
L

['a', 'c']

In [38]:
L[::-1]

['c', 'a']

## 2.  Numpy arrays

In [8]:
import numpy as np
a = np.arange(15)
print(a)
a

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

This is called an **array** and for now you can think of this as sequence with only one type.  You can do various sequence like things, with arrays, such as indexing by number and taking splices.

But we haven't yet seen what arrays really can do.   In addition to these sequence-like arrays there are also **2D arrays**, which are **tables of numbers**.  Here both row and column structure matter.

In [9]:
LL_list = list(range(15))
print(LL_list)
LL = np.array([LL_list[0:5],LL_list[5:10],LL_list[10:15]])
print(LL)
#print(LL[1])
#print(a)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


In [12]:
a[13]

13

In [43]:
LL[2,3]

13

In [15]:
LL[2,:]

array([10, 11, 12, 13, 14])

In [46]:
LL[2,0:2]

array([10, 11])

In [48]:
print(a)
a[2]

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


2

Why?  `a` is a 1D array

In [49]:
a_1d = np.arange(15)
print(a_1d)
a = a_1d.reshape(3, 5)
print(a)
b = np.arange(15).reshape(5, 3)
a = np.arange(15).reshape(3, 5)
print(b)
print(b.transpose())

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]
 [12 13 14]]
[[ 0  3  6  9 12]
 [ 1  4  7 10 13]
 [ 2  5  8 11 14]]


The need for 2D arrays is obvious if you've taken a linear algebra class.  They correspond to the mathematical object called a **matrix**.  One application for matrices is in solving systems of equations, but this really only scratches the surface.  A more fundamental application for a data scientist is they can represent all that is known about a single data set.  Generally each row represents an item (an individual or event in the data), and the entry in each column its value for a particular attribute. For example let's say each room in a hotel has lamps, tables, chairs, and beds, but in varying numbers.  We might represent the inventory of items in 5-room hotel with a 5x4 matrix (5 rows, 4 columns), as follows:

```
6  3 4 1
5  2 3 2
8  3 6 2
5  1 3 1
10 4 7 2
```

So the first row represents a room with 6 lamps, 3 tables, 4 chairs, and 1 bed.
Now if we represent the cost of each item as 1D cost array (or **vector**) to use the mathematical term,

```
40 175 90 450,
```

where the costs are ordered in the same way as our columns above: lamps costs, table costs, chair costs, and bed costs.  Then we can compute the per room costs as follows:

In [25]:
240 + 525 + 360 + 450

1575

In [26]:
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In [18]:
print(room_matrix)
print(cost_vector)
print(room_matrix.dot(cost_vector))

[[ 6  3  4  1]
 [ 5  2  3  2]
 [ 8  3  6  2]
 [ 5  1  3  1]
 [10  4  7  2]]
[ 40 175  90 450]
[1575 1720 2285 1095 2630]


The mathematical name for the `dot` method, which computes the per room cost, is **matrix multiplication**.

Arrays have many attributes.

In [50]:
print(a.shape)
print(a.ndim)
print(a.dtype.name)
print(a.itemsize)
print(a.size)
print(type(a))
print(np.ndarray)

(3, 5)
2
int64
8
15
<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [20]:
b = np.array([6, 7, 8])
print(b)
print(type(b))
b

[6 7 8]
<class 'numpy.ndarray'>


array([6, 7, 8])

A sequence of sequences can be used to define an array, but since the inner sequences are rows, they must all be the same length.

In [21]:
b = np.array( [ (1.5,2,3), (4,5,6) ] )
b

array([[1.5, 2. , 3. ],
       [4. , 5. , 6. ]])

Arrays can also contain complex numbers, but remember it takes two quantities to define a single complex number::

In [22]:
c = np.array( [ [1,2], [3,4] ], dtype=complex )
c

array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]])

A very convenient way to fill an array is to start with an array containing all 0's or 1's and then update the contents:

In [25]:
X = np.zeros( (3,4) )
X

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [28]:
Y = np.arange(12).reshape((3,4))
print(Y)
Y[2,3]

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


11

## Indexing

Indexing works as it does with ordinary Python sequences, except that now there are two dimensions to worry about

In [94]:
L= list(range(12))
print(L)
L[0:3] = [3,4]
print(L)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
[3, 4, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [51]:
X = np.zeros((3,4))

In [52]:
X[0,0]

0.0

In [53]:
X[1,2]

0.0

In [54]:
print(X)
X[2,0]

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


0.0

Assigment works as it does with Python sequences, except that it now works in two dimensions

In [55]:
X[1,2] = 3
X

array([[0., 0., 0., 0.],
       [0., 0., 3., 0.],
       [0., 0., 0., 0.]])

In [16]:
X = np.zeros( (3,4), dtype=int )
ctr = 0
(rows,cols) = X.shape
for i in range(rows):
    for j in range(cols):
        ctr += 1
        X[i,j] += ctr*10
X

array([[ 10,  20,  30,  40],
       [ 50,  60,  70,  80],
       [ 90, 100, 110, 120]])

In [35]:
type(X[1,2])

numpy.float64

In [38]:
import numpy as np
X = np.arange(10,130,10).reshape((3,4))
X

array([[ 10,  20,  30,  40],
       [ 50,  60,  70,  80],
       [ 90, 100, 110, 120]])

In [20]:
Y[:]

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [39]:

print(X)
print(X[2,3])
print(X[2,:])
print(X[:,2])
print(X[1:3,1:3])

[[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]
120
[ 90 100 110 120]
[ 30  70 110]
[[ 60  70]
 [100 110]]


In [5]:
X = np.arange(10,130,10).reshape((3,5))

ValueError: cannot reshape array of size 12 into shape (3,5)

The following expression gives the second row of `X`

```
X[1,:]
```

The following expression gives the 3rd column of `X`:

```
X[:,2]
```

Note that columns and rows are just 1D arrays; nothing identifies them as row vectors or column vectors.

As a convenience, `numpy` lets you skip the `:` when you want a row.  In other words, the next two expressions are synonyms.  Note that you **need** the `:` if you want a column.

In [35]:
print(X[1,:])
print(X[1])

[50 60 70 80]
[50 60 70 80]


In [39]:
print(X)
print(X[1,:], X[1,:].shape)
print(X[:,1])

[[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]
[50 60 70 80] (4,)
[ 20  60 100]


Next let's get part of a column or row.  The first two elements of the second column of `X`, followed by the first two elements of the second row of `X`:

In [40]:
print(X)
print(X[1,:2])
print(X[:2,1])

[[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]
[50 60]
[20 60]


Now let's get a 2x2 subarray, consisting of the upper left hand corner of `X`, followed by the central part of the last two rows.

In [41]:
print(X)
print(X[:2,:2])
print(X[1:,1:3])

[[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]
[[10 20]
 [50 60]]
[[ 60  70]
 [100 110]]


Array transposition is sometimes a nice way to get to the 2D array you eally want.  The transposition of an array `M` is called `M.T`, and the definition is that 

```
M.T[i,j] = M[j,i]
```

So if `M` is an `m` x `n` array, then `M.T` is an `n` x `m`  array.   Look at `X.T` and verify these definitions.  The `m`th row of `X` becomes the `m`th column of `X.T`. The `n`th column  of `X` becomes the `n`th row of `X.T`.

In [42]:
print(X)
print()
print(X.T)
print(X[1,2], X.T[2,1])

[[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]

[[ 10  50  90]
 [ 20  60 100]
 [ 30  70 110]
 [ 40  80 120]]
70 70


Another set of examples.  We'll explain the way `a` is created in a bit more detail below.  For now, let's just use it as an indexing example.  The point here is that `a` is a 1-dimensional array, which means that only one index is required, or allowed!

In [3]:
a = np.arange(10)**3
a

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729])

What to do to see that it's 1-dimensional

In [57]:
a.shape

(10,)

In [54]:
a[2]

8

In [9]:
a[2,0]

IndexError: too many indices for array

So indexing works much as it does with lists.

In [55]:
a[2:5]

array([ 8, 27, 64])

In th next cell we demonstrate assignment to a splice.  In normal Python the value that is assigned to a splice has to be a sequence.  But in numpy it is treated as an elementwise operation, so in the next  example, we assign the value -1000 to each of the positions in the splice.

A big difference between a 1-dimensional array and a list is that 1D arrays allow a more gener class of assignments.

In [58]:
print(a)
a[:6:2] = -1000    # equivalent to a[0:6:2] = -1000; from start to position 6, exclusive, set every 2nd element to -1000
print(a)

[  0   1   8  27  64 125 216 343 512 729]
[-1000     1 -1000    27 -1000   125   216   343   512   729]


In [ ]:
a[ : :-1]                                 # reversed a

There are also 3D arrays, which have a third dimension; each position along the third dimension defines a 2D array.

In [86]:
A =np.arange(24).reshape((2,3,4))
A

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

 In a 2D array, a single index slice is a row or column (a 1D array).  In a 3D array, a single index slice is a 2D array.

In [87]:
A[1,:,:]

array([[12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [88]:
A[:,:,3]

array([[ 3,  7, 11],
       [15, 19, 23]])

In [89]:
A[:,2,:]

array([[ 8,  9, 10, 11],
       [20, 21, 22, 23]])

We will not be using 3D arrays much, and they will not be featured on quizzes.  But it's nice to know they're out there.

## 3.  Elementwise arithmetic operations

In [27]:
import numpy as np
a = np.array( [20,30,40,50] ).reshape((2,2))
b = np.arange( 4 ).reshape((2,2))
print(a)
print(b)

[[20 30]
 [40 50]]
[[0 1]
 [2 3]]


In [41]:
a + b

array([[20, 31],
       [42, 53]])

In [42]:
a * b

array([[  0,  30],
       [ 80, 150]])

In [43]:
a - b

array([[20, 29],
       [38, 47]])

In the next example we illustrate **broadcasting**.  Elementwise operations are extended to apply between arrays and objects that are not arrays.  So for example, 2 * a (a is the array above) returns an array that contains all the elements of a multiplied by 2

In [58]:
print(a)
2 * a

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


array([[ 0,  2,  4,  6,  8],
       [10, 12, 14, 16, 18],
       [20, 22, 24, 26, 28]])

This works by "broadcasting" 2 into an array the same size as a and then doing elementwise mutliplication on the two arrays.
The next example applies two operations to `a`, first `sin`, then multiplication by 10.

You can multiply two arrays of the same shape together as we did with `a` and `b`
above, but you can't rely on broadcasting to figure out what to do with mismatched array sizes. In general, multiplying arrays of different sizes together will fail:

In [63]:
c = np.array([3,2,1,5,4])
print(a)
c * a

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


array([[ 0,  2,  2, 15, 16],
       [15, 12,  7, 40, 36],
       [30, 22, 12, 65, 56]])

The next example demonstrates the possibility of applying and broadcasting two operations to `a`, first `sin`, then multiplication by 10.

In [46]:
10*np.sin(a)

array([[ 9.12945251, -9.88031624],
       [ 7.4511316 , -2.62374854]])

The same thing works with Boolean tests, a point we will will return to below.

In [50]:
print(a)
Y = a < 35
print(Y)

[[20 30]
 [40 50]]
[[ True  True]
 [False False]]


In [49]:
a[Y]

array([20, 30])

In [51]:
print(a)
print(b)
a == b

[[20 30]
 [40 50]]
[[0 1]
 [2 3]]


array([[False, False],
       [False, False]])

In [52]:
if a == b:
    print("Hi!")

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [29]:
A = np.array( [[1,1],
            [0,1]] )
B = np.array( [[2,0],
            [3,4]] )

In [30]:
A * B  # elementwise product

array([[2, 0],
       [0, 4]])

## Fast and slow array creation

In [36]:
import numpy as np


# Slow
X = np.zeros( (5,4), dtype=int )
(rows,cols) = X.shape
for i in range(rows):
    for j in range(cols):
        X[i,j] = 10 * i + j
print(X)

# Fast
def f(i,j):
      return 10*i+j

# make a 5x4 array [20 elements] `b` using the function `f`
# where b[i,j] = f(i,j)
b = np.fromfunction(f,(5,4),dtype=int)
print(b)

[[ 0  1  2  3]
 [10 11 12 13]
 [20 21 22 23]
 [30 31 32 33]
 [40 41 42 43]]
[[ 0  1  2  3]
 [10 11 12 13]
 [20 21 22 23]
 [30 31 32 33]
 [40 41 42 43]]


## Elementwise Boolean operations, Masking

A very important extension of elementwise operations is the extension to Boolean tests.

Applying a Boolean test to an array returns an array of truth-values: Just as

```
3 + X
```

adds 3 to every element of array `X`, so

```
X > 2
```

returns an array of truth-values which tells us which elements of `X` are greater than 2.

In [32]:
import numpy as np
X = np.array([1,5,2,7,9,4,3,-6])
print(X)
print(X >= 3)

[ 1  5  2  7  9  4  3 -6]
[False  True False  True  True  True  True False]


What is the point of having such things around?  

Here it is:  Boolean arrays can be used to index other arrays.  They must be the same shape as the array
they index.   So in the simplest case:

In [16]:
print(X)
print(X[np.array([False,  True, False,  True,  True,  True,  True, False])])

[ 1  5  2  7  9  4  3 -6]
[5 7 9 4 3]


But now I can take advantage of the fact `X>=3` is a Boolean array to 

In [17]:
print(X)
print(X>=3)
print(X[X>=3])

[ 1  5  2  7  9  4  3 -6]
[False  True False  True  True  True  True False]
[5 7 9 4 3]


This returns exactly the members of X that are greater than or equal to 3.

We could do the same with a list comprehension, but the array computation above is much faster:

In [34]:
np.array([x for x in X if x >= 3])

array([5, 7, 9, 4, 3])

Boolean arrays can also be used to **count** the number of elements in an array that satisfy some constraint. Although `len` works on arrays and could be used, `sum` is faster.  The Boolean `True` is treated as 1 and the Boolean `False` as 0, so summing a Boolean array counts
the number of trues.  The following expression correctly counts the number of elements in `X` that are greater than or equal to 3.

In [15]:
sum(X>=3)

5

More on array operations and efficiency below.  It's a tricky subject.

The reason that `X[X>=3]` works has to do with a basioc fact of array indexing we haven't
really made explicit yet.  Any Boolean array Y of the right length can be used to index an array X if X and Y are the same length.  

A hokey but essential example.  Suppose we have an array of length 8 and we want to access the first, third, and seventh elements.  We can of course do it with a Boolean array:

In [23]:
X = np.arange(8) + 1
Y = np.array([True,False,True,False,False,False,True,False])
print(X)
X[Y]

[1 2 3 4 5 6 7 8]


array([1, 3, 7])

But there's a better way!

Cook up a sequence consisting of the **indices** of the first, third, and seventh elements, namely:

```
[0,2,6]
```

Now use that sequence as a **fancy** index on X:

In [25]:
X[[0,2,6]]

array([1, 3, 7])

Note the difference between what we just did and passing three different indices to the 1D array `X`,
which is an error.

In [26]:
X[0,2,6]

IndexError: too many indices for array

Fancy indexing also works on 2D arrays.

In [93]:
Y = np.arange(12).reshape((3,4))
print(Y)
print(Y[[0,2],[1,2]])

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[ 1 10]


Back to Boolean operations.  Boolean operations work more or less as expected on 2D arrays:

In [55]:
import numpy as np
y = np.arange(35).reshape(5,7)
y>3

array([[False, False, False, False,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True]], dtype=bool)

Note that this Boolean array preserves the 5x7 shape of `y`.

In [56]:
(y>3).shape

(5, 7)

But if we use it to index elements of `y`, the result has to be a 1D array:

In [57]:
y[y>3]

array([ 4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
       21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34])

Now suppose we have an array of data that has 4 columns, but we only want to keep the
first, second and fourth columns.  Here's a simple way to do that using
a Boolean array:

In [58]:
X = np.arange(8).reshape((2,4)) + 1
print(X)
print(X[:,np.array([True,True,False,True])])

[[1 2 3 4]
 [5 6 7 8]]
[[1 2 4]
 [5 6 8]]


Now Boolean arrays and Boolean masking is a great feature but it will occasionally cause you grief if you forget that something is an array and that a Boolean test will often return an array of truth values instead of a truth value.  

Python will return an error if you try to use an array where a Boolean value is expected. The most common error of this sort is in an `if`-clause.  

In [94]:
y = np.zeros((2,2))
# Hoping y is equivalent to False
if y:
    print('hi')

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

And the result of any Boolean test that can be interpreted as an elementwise operation will raise the same error:

In [66]:
if y == 0:
    print('hi')

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [96]:
y[0,1] = 2
y == 0

array([[ True, False],
       [ True,  True]])

This is what Python means by saying "The truth value of an array with more than one element is ambiguous."  Generally it's a mixture of Truths and falsities, like real life.

## Data in the form of arrays

Many python modules that provide data do so in array form.  As an example,
we load the famous **iris data set** due to Ronald Fisher, familiar to many who've had a statistics class, but also to many who've a computer science or machine learning class in which data analysis plays a role.

In [71]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

The name `data` has been set to a dictionary.

In [72]:
list(data.keys())

['data', 'target', 'target_names', 'DESCR', 'feature_names']

In [61]:
data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

The variable `features` has been set to `data['data']` and that is a `numpy` array:

In [39]:
print(features.shape)
print(features.ndim)
print(features.dtype.name)
print(features.size)
print(type(features))
print(features[:10])

(150, 4)
2
float64
600
<class 'numpy.ndarray'>
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


The last thing printed is the first 10 rows of `features`.  This array represents data
about irises.  Each row represents a different iris and gives 4 measurements for that 
exemplar.  So there are 150 iris exemplars; with 4 measurements for each, that's 600 items of
data in the array (`features.size`).  The data is used for classification studies.

Let's apply what we've just learnt to pick a certain subset of the data. 
The first number
in each row represents the **sepal length** of that particular iris exemplar (to be demonstrated below).

Find the flowers whose sepal length is exactly 5.

In [78]:
# 1D array: first column
first_col = features[:,0]
print(first_col == 5.0)

[False False False False  True False False  True False False False False
 False False False False False False False False False False False False
 False  True  True False False False False False False False False  True
 False False False False  True False False  True False False False False
 False  True False False False False False False False False False False
  True False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False  True False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False]


Then we use the Boolean array just constructed to index the rows, combined with a `:` to indicate we want all columns.

In [79]:
features[first_col == 5.0,:]

array([[5. , 3.6, 1.4, 0.2],
       [5. , 3.4, 1.5, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5. , 3.2, 1.2, 0.2],
       [5. , 3.5, 1.3, 0.3],
       [5. , 3.5, 1.6, 0.6],
       [5. , 3.3, 1.4, 0.2],
       [5. , 2. , 3.5, 1. ],
       [5. , 2.3, 3.3, 1. ]])

So there are only 10 plants out of 150 that have sepal lengths of exactly 5.

Of course we can do this all in one step, with the same result:

In [6]:
features[features[:,0] == 5.0,:]

array([[ 5. ,  3.6,  1.4,  0.2],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5. ,  3.2,  1.2,  0.2],
       [ 5. ,  3.5,  1.3,  0.3],
       [ 5. ,  3.5,  1.6,  0.6],
       [ 5. ,  3.3,  1.4,  0.2],
       [ 5. ,  2. ,  3.5,  1. ],
       [ 5. ,  2.3,  3.3,  1. ]])

Since we're indexing rows here, the `:` can also be left out:

In [25]:
features[features[:,0] == 5.0]

array([[ 5. ,  3.6,  1.4,  0.2],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5. ,  3.2,  1.2,  0.2],
       [ 5. ,  3.5,  1.3,  0.3],
       [ 5. ,  3.5,  1.6,  0.6],
       [ 5. ,  3.3,  1.4,  0.2],
       [ 5. ,  2. ,  3.5,  1. ],
       [ 5. ,  2.3,  3.3,  1. ]])

A more interesting use of Boolean array indexing is to find all the irises of a particular class, using the variable `target` defined when we loaded the data; `target` is an array containing the class of each iris in the data set:

In [76]:
print(type(target))
print(target.shape)
print(set(target))
# Grab the first rows of both arrays
print(features[0], target[0])
# Grab ro w indexed 90 from both arrays
print(features[90], target[90])

<class 'numpy.ndarray'>
(150,)
{0, 1, 2}
[5.1 3.5 1.4 0.2] 0
[5.5 2.6 4.4 1.2] 1


As the printouts indicate , `target` is a 1D array (a **vector**) of length 150, containing only the values 0,1, and 2.  These are the three classes to which an iris can belong.  There are exactly as many entries in the `target` array as there are rows in the `features` array. For any iris, its row-index in the `features` array is the index of its class in the `target`.  Above we printed the features and target class for irises 0 and 90.

We can take advantage of this structure to find all the irises of class 1 very efficiently.
Note that `target == 1` is a Boolean array of length 150.

In [27]:
target==1

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

To find the irises of class 1, we simply use that Boolean array to specify the rows
we want in the `features` array:

In [50]:
features[target==1,:].shape

(50, 4)

There are 50 rows whose class is `1`; you can verify for yourselves that are also 50 rows for classes 2 and 3.  This is a balanced data set.

Going back to the dictionary we loaded when we loaded the datset, we can
find the meanings of our 4 columns:

In [17]:
data['feature_names']

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

So our 4 columns have numerical data on these attributes. The column indexed 0 contains the sepal length measurement for 1 flower in our sample.  The column indexed 1 contains sepal width measurements. And so on.

We know there are three classes (0, 1, and 2). We can look up their names under the `'target_names'` key of the dictionary we got when we loaded the data.  

In [18]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], 
      dtype='|S10')

So class 0 represent the iris species `setosa`, and so on.

Here's what the entire target array looks like:

In [19]:
data['target']

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Find the irises whose species is 1, and whose sepal length is greater than 5.

In [75]:
# New smaller data table consisting of those flowers of species 1.
species1 = features[target == 1,:]
print(len(species1))
species1[species1[:,0]>=5.0,:]
print(len(species1[species1[:,0]>=5.0,:]))

50
49


In [87]:
print(species1[:,0]>=5.0)
sum(species1[:,0]>=5.0)

[ True  True  True  True  True  True  True False  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True]


49

Now suppose we wanted to just see the **sepal lengths** (column 0) of the target class 1 irises, and compare them to the sepal lengths of the target class 0 irises, to see if sepal lengths provided a good way of telling the two classes apart.  We could do:

In [51]:
print(features[target==1,0])
print(features[target==0,0])

[7.  6.4 6.9 5.5 6.5 5.7 6.3 4.9 6.6 5.2 5.  5.9 6.  6.1 5.6 6.7 5.6 5.8
 6.2 5.6 5.9 6.1 6.3 6.1 6.4 6.6 6.8 6.7 6.  5.7 5.5 5.5 5.8 6.  5.4 6.
 6.7 6.3 5.6 5.5 5.5 6.1 5.8 5.  5.6 5.7 5.7 6.2 5.1 5.7]
[5.1 4.9 4.7 4.6 5.  5.4 4.6 5.  4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.  5.  5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.
 5.5 4.9 4.4 5.1 5.  4.5 4.4 5.  5.1 4.8 5.1 4.6 5.3 5. ]


And though there is some overlap, we see that there is a tendency for the column-0
value be higher in target 1 than it is target class 0.

## Cosine

One way of measuring the similarity between two arrays `x` and `y`. is to take their cosine.  The name for this similarity measure is quite appropriate.  The cosine of `x` and `y` is a measure of the geometric cosine of the angle between them.  It is 1 when that angle is 0 (i.e., `x` and `y` are identical in direction), and -1 when the vectors point in opposite directions, and 0 when the vectors are orthogonal (the do not share a component in any direction).  It is computed by taking the dot product of the unit vectors pointing in the same direction as `x` and `y`.  To do this we divide `x` and `y` by their Euclidean length (`LA.norm` in the code below) and take the dot product of the results.

In [64]:
import numpy as np
import numpy.linalg as LA
x = np.array([3,4])
y = np.array([4,3])

def find_unit_vector(x):
    return x/LA.norm(x,ord=2)

def cosine (x,y):
    return (find_unit_vector(x)).dot(find_unit_vector(y))

The maximum possible value cosine can have for two vectors is `1.0`.  The maximum possible value is achieved when you compare a vector `x` with itself `x`.  So nothing can be more similar to `x` than `x` is to itself.  Whew.  That's reassuring.

In [65]:
cosine(x,x)

1.0

In [66]:
cosine(x,-x)

-1.0

In [67]:
cosine(x,y)

0.96

In [68]:
LA.norm(find_unit_vector(x))

1.0

## Why use arrays?

Arrays provide all kinds of indexing convenience for accessing data that
is genuinely **tabular**.  Tabular data is data that has good reason for being represented in
a table;  that is, there is good reason to think of the rows as a unit, and good reason
to think of the columns as units.  And therefore, there is good reason to think we might want
to apply some computational operations to columns and rows.  We'll explore those ideas a bit more
in a future assignment.  

All kinds of data is fundamentally tabular, as we'll see.  But although that's very important,
it's probably not the main reason why so many scientists, engineers, and data analysts use `numpy`
and other programming tools like `matlab` and `R` that provide tables, or **matrices** (the mathematical
name for a similar concept), as a fundamental data structure.  The main reason is computational efficiency.

Having a single data structure with predictable types and elementwise basic arithmetic operations makes
it possible to write very efficient ways of performing more complex operations.  The above code snippet
demonstrates this by providing a timing comparison between two ways of computing teh sum of the
squares between 0 and 999 inclusive; the first is used in computing the variable `normal_py_sec`, is the ordinary Python way of computing such a sum:  Cook up a container of theose integers, square each of them and sum the result.
The second, stored in the varibale `np_sec`, uses numpy arrays.  There is actually a very important
mathematical operation **dot product** that multiplies all the corresponding elements of two vectors (elemntwise) and
sums the result.  Using this `precompiled` fast math operation we speed the code up by a factor of about
 80 on my machine, and your results, on your various home machines will be different, but in the same ballpark.
 
Note that numpy arrays in and of themselves don't help.  In the code snippet below, we multiply the `na` array times itself and sum the result.  This code is actually *slower* than the basic Python code, because it pays the overhead of creating the `numpy` array but doesn't exploit it by using the proper optimized operation.  The lesson here is that coding so as to  take maximum advantage of the efficiencies of `numpy` can be tricky, but if you have a good reason to be efficient, it's definitely worth some experimentation. That starts with knowing about the kinds of builtin array specific functions `numpy` offers.

In [3]:
import timeit
import numpy as np

py_sec = timeit.timeit('sum(x*x for x in range(1000))',
                              number=10000)

np_sec = timeit.timeit('na.dot(na)',
                            setup="import numpy as np; na=np.arange(1000)",
                            number=10000)

print(f"Normal Python: {py_sec} sec")

print(f"NumPy: {np_sec} sec")

Normal Python: 0.9874542279867455 sec
NumPy: 0.018804321996867657 sec


In [31]:
naive_np_sec = timeit.timeit('sum(na*na)',
                             setup="import numpy as np; na=np.arange(1000)",
                             number=10000)
print(("Naive NumPy: %f sec" % naive_np_sec))

Naive NumPy: 0.943414 sec
